### PR TITLE
Sort by a new 'priority' parameter

### PR DESCRIPTION
**Pull request by tomap**
Bump minor, because there could be a minor change in behavior
if some user already have a priority parameter
on their posts

Based upon the work of #6

### CODE REVIEW

**Comment by YunYouJun** [comment]
I've been looking forward to this feature for a long time.
The earlier merged, the better.

---

By the way, I found some theme such as [hexo-theme-next](https://github.com/theme-next/hexo-theme-next/), [hexo-theme-melody](https://github.com/Molunerfinn/hexo-theme-melody) etc.
They had added `sticky` style.
Maybe use `sticky` replace `priority` is ok?

**Comment by stevenjoezhang** [comment]
The parameter `config.index_generator.order_by` is ignored

**Comment by weyusi** [comment]
> The parameter `config.index_generator.order_by` is ignored

That's the intention.
```
note that if your posts have a `priority` property, it will be used to order your post taking precedence over `order_by`.
```

**Comment by stevenjoezhang** [comment]
> That's the intention.

> note that if your posts have a `priority` property, it will be used to order your post taking precedence over `order_by`.

But how about posts with same priority? It does not match the description.

**Comment by tomap** [comment]
well, I just looked at the code again, and it is not what I intended...
I was trying to reproduce the work of https://github.com/hexojs/hexo-generator-index/pull/6
But what we have here is something that will break the parameter config.index_generator.order_by in all cases, which is not good

I believe the configuration config.index_generator.order_by can be used to order by priority, if a priority has been defined on all posts
What might be missing is a was to order by priority first, if there is a priority defined, it goes on top
If not, the config.index_generator.order_by is used
I'm not sure how to code that :)

**Comment by stevenjoezhang** [comment]
Yes, the issue is: When two or more posts have the same `priority`, the sorting should fall back to `config.index_generator.order_by`, but not
```js
if(a.priority == b.priority) return b.date - a.date;
```

**Comment by stevenjoezhang** [comment]
> But what we have here is something that will break the parameter config.index_generator.order_by in all cases, which is not good

@tomap I think you are talking about the stability of the sorting algorithm. Node.js 12+ uses stable sorting algorithms by default: https://v8.dev/features/stable-sort . Therefore, you can sort `locals.posts` by `config.index_generator.order_by` (e.g. '-date') first, and then sort `locals.posts.data` according to the value of `sticky` or `priority` parameter. For older Nodejs versions, `lodash` is required to implement this method.

File: lib/generator.js [hunk 1]
```diff
@@ -4,7 +4,21 @@ const pagination = require('hexo-pagination');
...
+      }
+      else if(a.priority && !b.priority) {
+          return -1;
+      }
+      else if(!a.priority && b.priority) {
+          return 1;
+      }
+      else return b.date - a.date;
+  });
+
```
Comment (stevenjoezhang) [inline lib/generator.js]:
```suggestion
  const posts = locals.posts;
  posts.data.sort((a, b) => (b.sticky || 0) - (a.sticky || 0));
  
```

**Comment by curbengh** [comment]
@stevenjoezhang 

can you create another PR with your patch? This PR has been inactive for a while...

**Comment by SukkaW** [comment]
cc @stevenjoezhang

**Comment by stevenjoezhang** [comment]
@curbengh @SukkaW done https://github.com/hexojs/hexo-generator-index/pull/51

**Pull request closed without merging**